src/qs8-vadd/gen/qs8-vadd-minmax-hvx-u128.c

```diff
@@ -0,0 +1,134 @@
+//
+// This source code is licensed under the BSD-style license found in the
+// LICENSE file in the root directory of this source tree.
+
+#include <assert.h>
+
+#include <hvx_hexagon_protos.h>
+#include <hexagon_protos.h>
+#include <hexagon_types.h>
+
+#include <xnnpack/vbinary.h>
+#include <xnnpack/intrinsics-polyfill.h>
+
+void xnn_qs8_vadd_minmax_ukernel__hvx_u128(
+    size_t batch,
+    const int8_t* input_a,
+    const int8_t* input_b,
+    int8_t* output,
+    const union xnn_qs8_add_minmax_params params[restrict XNN_MIN_ELEMENTS(1)]) XNN_OOB_READS
+{
+  assert(batch != 0);
+  assert(batch % sizeof(int8_t) == 0);
+  assert(input_a != NULL);
+  assert(input_b != NULL);
+  assert(output != NULL);
+
+  const HVX_Vector vbias = Q6_V_vsplat_R(*((int32_t *) &params->hvx.bias));
+  const HVX_Vector va_multiplier = Q6_V_vsplat_R(*((int32_t *) &params->hvx.a_multiplier));
+  const HVX_Vector vb_multiplier = Q6_V_vsplat_R(*((int32_t *) &params->hvx.b_multiplier));
+  const int32_t shift = params->hvx.shift;
+  const HVX_Vector voutput_zero_point = Q6_Vh_vsplat_R(*((int16_t *) &params->hvx.output_zero_point));
+  const HVX_Vector voutput_min = Q6_Vb_vsplat_R(*((int8_t *) &params->hvx.output_min));
+  const HVX_Vector voutput_max = Q6_Vb_vsplat_R(*((int8_t *) &params->hvx.output_max));
+  int8_t* ptr_o = output;
+
+  for (; batch >= 128 * sizeof(int8_t); batch -= 128 * sizeof(int8_t)) {
+    HVX_Vector va0 = *((HVX_UVector*)input_a);
+    HVX_Vector vb0 = *((HVX_UVector*)input_b);
+    input_a += 128;
+    input_b += 128;
+
+    // unpack: 8 bit to 16 bit
+    HVX_VectorPair va0_i16 = Q6_Wh_vsxt_Vb(va0);
+    HVX_Vector va0_i16_even = Q6_V_lo_W(va0_i16);
+    HVX_Vector va0_i16_odd = Q6_V_hi_W(va0_i16);
+
+    HVX_VectorPair vb0_i16 = Q6_Wh_vsxt_Vb(vb0);
+    HVX_Vector vb0_i16_even = Q6_V_lo_W(vb0_i16);
+    HVX_Vector vb0_i16_odd = Q6_V_hi_W(vb0_i16);
+
+    // vacc = va * va_multiplier + vb * vb_multiplier with expanding to 32 bit
+    HVX_VectorPair va0_mul_even = Q6_Vw_vmpyi_VwVh(va_multiplier, va0_i16_even);
+    HVX_VectorPair va0_mul_odd = Q6_Vw_vmpyi_VwVh(va_multiplier, va0_i16_odd);
+    HVX_VectorPair vb0_mul_even = Q6_Vw_vmpyi_VwVh(vb_multiplier, vb0_i16_even);
+    HVX_VectorPair vb0_mul_odd = Q6_Vw_vmpyi_VwVh(vb_multiplier, vb0_i16_odd);
+
+    HVX_Vector vacc0_even = Q6_Vw_vadd_VwVw(Q6_V_lo_W(va0_mul_even), Q6_V_lo_W(vb0_mul_even));
+    HVX_Vector vacc0_odd = Q6_Vw_vadd_VwVw(Q6_V_lo_W(va0_mul_odd), Q6_V_lo_W(vb0_mul_odd));
+    HVX_Vector vacc1_even = Q6_Vw_vadd_VwVw(Q6_V_hi_W(va0_mul_even), Q6_V_hi_W(vb0_mul_even));
+    HVX_Vector vacc1_odd = Q6_Vw_vadd_VwVw(Q6_V_hi_W(va0_mul_odd), Q6_V_hi_W(vb0_mul_odd));
+
+    // vacc = vbias + vacc
+    vacc0_even = Q6_Vw_vadd_VwVw(vbias, vacc0_even);
+    vacc0_odd = Q6_Vw_vadd_VwVw(vbias, vacc0_odd);
+    vacc1_even = Q6_Vw_vadd_VwVw(vbias, vacc1_even);
+    vacc1_odd = Q6_Vw_vadd_VwVw(vbias, vacc1_odd);
+
+    // right shift, lower to 16 bit, add output_zero_point
```

Comment:
Can you use `Q6_Vh_vasr_VwVwR_sat` *first* by 16 bits, and then shift by the rest after narrowing to 16-bit? That would be slightly better if so...

Reply:
Also: I think `Q6_Vh_vasr_VwVwR_sat` shifting by 0 or 16 is equivalent to `vshuffe` or `vshuffo` respectively (I can't remember which is which or if I even have the right instructions...).